# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/res/cs_fast_linear_gradient.glsl
# Commit: e38dce4a5b06
# Full Hash: e38dce4a5b06869dd4a4d753f41f7594a8cddfb6
# Author: Nicolas Silva <nsilva@mozilla.com>
# Date: 2021-04-22 21:31:57
# Regressor Bug: 1702228
# File Overlap Count: 2
# Description:
#   Bug 1702228 - Cache linear gradients by default. r=gw
#   
#   This patch breaks linear gradients into two primitives:
#    - LinearGradient is always rendered via the brush shader. It is only used with SWGL.
#    - CachedLinearGradient is always rendered via a cached render task and the brush image shader. Its implementation is very similar to conic and radial gradients, with the addition of a fast-path for axis aligned gradients with two stops.
# ==============================================================================

diff -r b3f8befa510f -r e38dce4a5b06 gfx/wr/webrender/res/cs_fast_linear_gradient.glsl
--- a/gfx/wr/webrender/res/cs_fast_linear_gradient.glsl	Thu Apr 22 10:34:54 2021 +0000
+++ b/gfx/wr/webrender/res/cs_fast_linear_gradient.glsl	Thu Apr 22 10:34:54 2021 +0000
@@ -5,52 +5,28 @@
 #include shared
 
 varying float vPos;
-flat varying vec4 vStops;
 flat varying vec4 vColor0;
 flat varying vec4 vColor1;
-flat varying vec4 vColor2;
-flat varying vec4 vColor3;
 
 #ifdef WR_VERTEX_SHADER
 
 PER_INSTANCE in vec4 aTaskRect;
-PER_INSTANCE in float aAxisSelect;
-PER_INSTANCE in vec4 aStops;
 PER_INSTANCE in vec4 aColor0;
 PER_INSTANCE in vec4 aColor1;
-PER_INSTANCE in vec4 aColor2;
-PER_INSTANCE in vec4 aColor3;
-PER_INSTANCE in vec2 aStartStop;
+PER_INSTANCE in float aAxisSelect;
 
 void main(void) {
-    vPos = mix(aStartStop.x, aStartStop.y, mix(aPosition.x, aPosition.y, aAxisSelect));
+    vPos = mix(0.0, 1.0, mix(aPosition.x, aPosition.y, aAxisSelect));
 
-    vStops = aStops;
     vColor0 = aColor0;
     vColor1 = aColor1;
-    vColor2 = aColor2;
-    vColor3 = aColor3;
 
     gl_Position = uTransform * vec4(aTaskRect.xy + aTaskRect.zw * aPosition.xy, 0.0, 1.0);
 }
 #endif
 
 #ifdef WR_FRAGMENT_SHADER
-float linear_step(float edge0, float edge1, float x) {
-    if (edge0 >= edge1) {
-        return 0.0;
-    }
-
-    return clamp((x - edge0) / (edge1 - edge0), 0.0, 1.0);
-}
-
 void main(void) {
-    vec4 color = vColor0;
-
-    color = mix(color, vColor1, linear_step(vStops.x, vStops.y, vPos));
-    color = mix(color, vColor2, linear_step(vStops.y, vStops.z, vPos));
-    color = mix(color, vColor3, linear_step(vStops.z, vStops.w, vPos));
-
-    oFragColor = color;
+    oFragColor = mix(vColor0, vColor1, vPos);
 }
 #endif